ARROW-1476: [JAVA] Implement Final ValueVector Updates

siddharthteotia opened the pull request:
cc @jacques-n , @BryanCutler , @icexelloss 

bug fixes and clean up done as part of testing in Dremio -- this is the 3rd patch I was referring to

wesm commented:
I just rebased this after ARROW-1717 was merged

wesm commented:
Can we try to fix up the checkstyle warnings as part of this process as well? There seems to be a bunch of indentation problems etc.

wesm commented:
We'll want to rebase the vector refactor branch after this is merged as some of the tests are failing due to reasons that have been fixed in master

siddharthteotia commented:
Yes @wesm , we need to rebase this branch before merging into master.  I have addressed indentation and line length issues.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -543,21 +533,28 @@ public void loadFieldBuffers(ArrowFieldNode fieldNode, List<ArrowBuf> ownBuffers
     */
    public List<ArrowBuf> getFieldBuffers() {
```
What's the difference of this function and 

```
public ArrowBuf[] getBuffers(boolean clear)
```
?

icexelloss replied:
@siddharthteotia What's your thoughts on this?

icexelloss replied:
Ping @siddharthteotia, can you explain on the difference between the two methods and how do we need both as public API?

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -430,11 +441,11 @@ private void allocateValidityBuffer(final long size) {
    }
 
-   protected void reallocValueBuffer() {
+   public void reallocDataBuffer() {
```
Docstring?

siddharthteotia replied:
Done.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -615,8 +632,10 @@ public void transferTo(BaseNullableVariableWidthVector target){
-      target.valueCount = valueCount;
-      target.setLastSet(lastSet);
+      target.setLastSet(this.lastSet);
+      if (this.valueCount > 0) {
```
Why do we need to check valueCount > 0?

siddharthteotia replied:
If we haven't done setValueCount() on the source vector before invoking transfer(), then valueCount is essentially 0. Doing target.setValueCount(this.valueCount) will corrupt the target vector if this.valueCount is 0.

icexelloss replied:
Ok.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -836,17 +858,259 @@ public int getLastSet() {
     * @param index position of the element.
     */
+   @Override
    public void setIndexDefined(int index) {
```
It seems naming conversion wise, all methods that handles allocates new buffers are called "xxxSafe". Does
```
public void setValiditySafe(int index, boolean valid)
```

Makes more sense for this function?

siddharthteotia replied:
Not quite sure. Since this has always been setIndexDefined(), I just left it as is

icexelloss replied:
Ok

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -836,17 +858,259 @@ public int getLastSet() {
+    * @param index   position of the element to set
+    * @param length  length of the element
+    */
+   public void setValueLengthSafe(int index, int length) {
```
I am curious what's the use case of this API, i.e, when do we want to just change the length?

icexelloss replied:
@siddharthteotia what's your thought on this?

icexelloss replied:
Ping @siddharthteotia , can you explain why do we need this as a public API? When would a user want to just sent length for an element?

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/GenerateSampleData.java:
```diff
@@ -0,0 +1,337 @@
+/*******************************************************************************
```
Don't know if it's github issue, is the license format correct?

siddharthteotia replied:
Corrected. Not a github issue.

icexelloss replied:
@siddharthteotia The license format still doesn't look correct.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/GenerateSampleData.java:
```diff
@@ -0,0 +1,337 @@
+ * provided generateTestData(now deprecated) API to populate the vector
+ * with sample data. This class should be used for that purpose.
+ */
+public class GenerateSampleData {
```
Should this class be under test or tools?

siddharthteotia replied:
Ideally under tools but for some weird reason, I wasn't able to resolve the dependency w.r.t arrow tools in Dremio -- I was definitely missing something but for saving time just put it here.

icexelloss replied:
I prefer if we can move this out of vector public class. If we don't have time to do it here, can we track this somewhere so we don't forget to move it before release?

icexelloss replied:
Ping @siddharthteotia. Looks like this class is not used anywhere. Do we need this?

siddharthteotia replied:
We use this in Dremio

icexelloss replied:
If it's only used for Dremio testing, does it make more sense to stay in Dremio codebase?

siddharthteotia replied:
This is not something new. Vector classes have had the API generateTestData() from the beginning. As part of refactoring, we decided to move it out of the main API of vectors and provide a separate API for populating vectors with dummy data.

siddharthteotia replied:
The old generateTestData() API in vectors is deprecated and the class description implies that any users of the former should now be using GenerateSampleData.

icexelloss replied:
I see. It makes more sense now, thanks for the explanation. Let's keep it then.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/NullableBitVector.java:
```diff
@@ -163,7 +163,7 @@ private ArrowBuf splitAndTransferBuffer(int startIndex, int length,
             if (destBuffer != null) {
                destBuffer.release();
             }
-            destBuffer = destBuffer.slice(firstByteSource, byteSizeTarget);
```
Are there unit tests for this kind of stuff?

siddharthteotia replied:
Yes, we have unit tests for splitAndTransfer of bit vector (validity buffer). Also we have unit tests  for split and transfer of other vectors (nullable fixed, var width and complex) and we exercise all the conditions.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/NullableBitVector.java:
```diff
@@ -279,6 +279,8 @@ public Boolean getObject(int index) {
    public void copyFrom(int fromIndex, int thisIndex, NullableBitVector from) {
       if (from.isSet(fromIndex) != 0) {
```
How about something like 

```
BitVectorHelper.setValidityBit(validityBuffer, thisIndex, BitVectorHelper.getValidity(from, fromIndex));
```
?

icexelloss replied:
I am also a bit confused about 

```
set(thisIndex, ...) 
```
vs
```
BitVectorHelper.setValidityBit(validityBuffer, thisIndex, ...)
```

What's the difference?

icexelloss replied:
Actually, nvm. It seems all `copyFrom` follows the same pattern.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/NullableBitVector.java:
```diff
@@ -279,6 +279,8 @@ public Boolean getObject(int index) {
       if (from.isSet(fromIndex) != 0) {
          set(thisIndex, from.get(fromIndex));
+      } else {
+         BitVectorHelper.setValidityBit(validityBuffer, thisIndex, 0);
```
Do we have a unit test for this?

siddharthteotia replied:
We have some copy() related unit tests but in any case, I am planning to write more unit tests in the next couple of weeks to get overall good coverage. I don't think so I can get to them now because of time constraints. 

Note that behavior of copyFrom() hasn't changed.

icexelloss replied:
Ok

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/NullableVarBinaryVector.java:
```diff
@@ -152,12 +135,9 @@ public void get(int index, NullableVarBinaryHolder holder){
       holder.isSet = 1;
-      holder.start = startOffset;
-      holder.end = dataLength;
+      holder.start = getstartOffset(index);
```
I think "offset" is well defined in format, so this method can probably just be:

```
getOffset(int index)
```

icexelloss replied:
or just:
```
offsetBuffer.getInt(index * OFFSET_WIDTH)
``

Probably it consist with `holder.end = ...`

siddharthteotia replied:
startOffset implies the starting point of VARCHAR element in the data buffer.

icexelloss replied:
Ok

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/complex/BaseRepeatedValueVector.java:
```diff
@@ -183,8 +183,13 @@ public void clear() {
+      buffers = new ArrowBuf[0];
+    } else {
+      buffers = ObjectArrays.concat(new ArrowBuf[]{offsetBuffer}, vector.getBuffers(false),
+              ArrowBuf.class);
```
Indentation seems incorrect.

siddharthteotia replied:
Done.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/complex/FixedSizeListVector.java:
```diff
@@ -282,7 +286,14 @@ public void clear() {
+      buffers = new ArrowBuf[0];
+    } else {
+      buffers = ObjectArrays.concat(new ArrowBuf[]{validityBuffer}, vector.getBuffers(false),
+              ArrowBuf.class);
```
Indentation seems incorrect.

siddharthteotia replied:
Done.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java:
```diff
@@ -128,17 +128,25 @@ public void loadFieldBuffers(ArrowFieldNode fieldNode, List<ArrowBuf> ownBuffers
     return result;
   }
 
+  private void setReaderAndWriterIndex() {
```
Can this be lifted to base class?

siddharthteotia replied:
Probably but I thought BaseRepeatedValueVector on itself is of no use -- at least doesn't represent a materialized field. Moreover, even if we move to base class, we still have to set reader and writer index for validity buffer in sub class since the validity buffer is a property of list vector and not BaseRepeatedValueVector.

icexelloss replied:
OK

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java:
```diff
@@ -436,8 +450,15 @@ public void clear() {
+      buffers = new ArrowBuf[0];
+    } else {
+      buffers = ObjectArrays.concat(new ArrowBuf[]{offsetBuffer},
+              ObjectArrays.concat(new ArrowBuf[] {validityBuffer},
```
Indentation

siddharthteotia replied:
Done.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java:
```diff
@@ -491,8 +512,22 @@ public int getNullCount() {
+    return Math.min(getValidityBufferValueCapacity(), super.getValueCapacity());
+  }
+
+  public int getValidityAndOffsetValueCapacity() {
```
Why do we need this method? Can valueCapacity and validityCapacity mismatch?

siddharthteotia replied:
validity and offset move together regardless of the buffers in underlying data vector of list vector so functions like setNotNull(), startNewValue(), setValueCount() which are essentially safe methods need to determine if the inner buffers of list vector need to be reallocated or not. This decision can't be made based on getValueCapacity() since that also accounts for the value capacity of data vector. I think by mistake it is public. that needs to be corrected.

icexelloss replied:
I see. Yeah let's make this private.

icexelloss commented:
@siddharthteotia This looks good. I left some comments.

One thing is that it seems this patch has quite bit of bug fixes but not many new tests. Should we add tests for those bug fixes?

siddharthteotia commented:
I rebased this patch on master and addressed review comments.

wesm commented:
Hm, I wonder what's going on with the merge conflicts?

siddharthteotia commented:
The conflicts have all been resolved in my local java-vector-refactor branch. The build and tests are clean. After doing "git pull --rebase upstream master" and addressing review comments, I did a force update of the branch for this PR? 

Everything is clean locally, not sure where I went wrong?

wesm commented:
Are you sure you rebased on apache/master? Like

`git rebase apache/master"

(here apache is git@github.com:apache/arrow.git and I have just run `git fetch apache`)

wesm commented:
Oh I see. The issue is the base branch of this PR. The rebase went fine

```
* 635ca770 2017-11-14 | address review comments (siddharthteotia/ARROW-1476) [siddharth]
* 343f47bd 2017-11-13 | rebase on master, fix tests [siddharth]
* 102ff972 2017-11-13 | indentation and line length [siddharth]
* 2027f0bb 2017-11-13 | ARROW-1476: [JAVA] Implement Final ValueVector Updates [siddharth]
* 0de6e697 2017-11-07 | ARROW-1717: [Java] Refactor JsonReader for new class hierarchy and fix integration tests [Li Jin]
* 4509c294 2017-10-16 | ARROW-1474:[JAVA] ValueVector hierarchy (Implementation Phase 2) [siddharth]
* 6493ebf8 2017-10-14 | ARROW-1473: ValueVector new hierarchy prototype (implementation phase 1) [Siddharth Teotia]
* b18bbeba 2017-11-13 | ARROW-1371: [Website] Add "Powered By" page to the website (wesm/master, apache/master, master) [Korn, Uwe]
* 8f2d1525 2017-11-13 | ARROW-1802: [GLib] Support arrow-gpu [Kouhei Sutou]
* 6f8e2873 2017-11-13 | ARROW-1743: [Python] Avoid non-array writeable-flag check [Licht-T]
* 4a33bad5 2017-11-12 | ARROW-1767: [C++] Support file reads and writes over 2GB on Windows (siddharthteotia/master, origin/master, origin/HEAD) [Licht-T]
* e8331f46 2017-11-12 | ARROW-1794: [C++/Python] Rename DecimalArray to Decimal128Array [Phillip Cloud]
* 7adadd8e 2017-11-12 | ARROW-1763: [Python] Implement __hash__ for DataType [Wes McKinney]
```

wesm commented:
Just changed the base

wesm commented:
OK, builds are running. I can take care of merging to master after the builds pass. @icexelloss @BryanCutler @jacques-n any last words before we pull the lever in an hour or two?

siddharthteotia commented:
Thanks @wesm for changing the base branch.

wesm commented:
The Java build is failing -- seems there is still quite a bit of multiple-of-3 indentation. Can you fix?

siddharthteotia commented:
I am fixing the multiple-of-3 indentation. I thought mvn clean install will fail the build locally if checkstyle problems are there. Is there a deterministic way to test this locally 

Looks like there is an integration test failure as well.

Incompatible files
index: 0, length: 1 (expected: range(0, 0))
19:50:30.037 [main] ERROR org.apache.arrow.tools.Integration - Incompatible files
java.lang.IndexOutOfBoundsException: index: 0, length: 1 (expected: range(0, 0))
	at io.netty.buffer.ArrowBuf.checkIndexD(ArrowBuf.java:138)
	at io.netty.buffer.ArrowBuf.chk(ArrowBuf.java:161)
	at io.netty.buffer.ArrowBuf.getByte(ArrowBuf.java:781)
	at org.apache.arrow.vector.BaseNullableFixedWidthVector.isSet(BaseNullableFixedWidthVector.java:799)
	at org.apache.arrow.vector.NullableDecimalVector.getObject(NullableDecimalVector.java:139)
	at org.apache.arrow.vector.NullableDecimalVector.getObject(NullableDecimalVector.java:39)
	at org.apache.arrow.vector.util.Validator.compareFieldVectors(Validator.java:122)
	at org.apache.arrow.vector.util.Validator.compareVectorSchemaRoot(Validator.java:101)
	at org.apache.arrow.tools.Integration$Command$3.execute(Integration.java:201)
	at org.apache.arrow.tools.Integration.run(Integration.java:114)
	at org.apache.arrow.tools.Integration.main(Integration.java:65)
	Suppressed: java.lang.IllegalStateException: Memory was leaked by query. Memory leaked: (4644)

 @icexelloss , is this the same integration test you were referring to in "refactor JsonReader PR?

icexelloss commented:
No, the integration test I saw is something else - the one I refer to is fixed by zeroing out validity bits in json reader.

siddharthteotia commented:
@icexelloss , I haven't dealt with integration tests yet. So this might be a dumb question. How do I run/debug an integration test locally?

icexelloss commented:
@siddharthteotia There is a README file under arrow/integration that should get your started.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/NullableBitVector.java:
```diff
@@ -0,0 +1,504 @@
+/*******************************************************************************
```
License format seems incorrect.

icexelloss commented:
For check style, I ran
```
mvn package checkstyle:check
```
under arrow/java. Not sure if there is a better way.

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/BaseValueVector.java:
```diff
@@ -122,5 +123,49 @@ public static boolean checkBufRefs(final ValueVector vv) {
+
+  protected void compareTypes(BaseValueVector target, String caller) {
+    if (this.getMinorType() != target.getMinorType()) {
+      throw new UnsupportedOperationException(caller + " should have vectors of exact same type");
```
was this in the API before, if not maybe it would be better to  just return a boolean and the caller can throw an exception?

icexelloss replied:
Since it's a protected method. I don't think it matters too much. I am ok with it as is.

siddharthteotia commented:
Thanks @icexelloss . The expectation is to have absolutely no checkstyle warnings? I can see several warnings for files that haven't been changes as part of refactoring patches.

siddharthteotia commented:
It looks like the integration test is failing because we need to make code changes in C++ that interprets the JSON and loads the vector. Since the vectors no longer have inner buffers, we had to change the JAVA JsonFileReader code in JAVA that parses the JSON and appropriately allocates the inner buffers (validity, offset, data) of the target vector and fills them with data.

We need to make similar changes in C++ code that parses a JSon file and loads/builds the vector. It looks like JAVA side of integration test is receiving a vector whose buffer has not been allocated.

I am digging but it will be very helpful if someone from C++ side can extend help in any manner.

wesm commented:
"Since the vectors no longer have inner vectors" I'm not familiar with this terminology -- the JSON format contains a vector along with its children (are these the inner vectors?). Let me know where the problem is occurring and I can take a look

wesm commented:
Does the JSON format need to change somehow?

siddharthteotia commented:
The Nullable scalars (fixed width and var width)

A NullableIntVector had two inner vectors -- BitVector and IntVector and we used to delegate the operations to these inner vectors. Now instead of the inner vectors, we just have two ArrowBufs that manage the validity and data respectively. 

The Json format doesn't need to change. However, the code that parses Json and writes to vector or vice-versa had to be changed in JAVA because it was written with assumption that vectors have inner vectors that can be directly operated on. It was changed to directly work with inner buffers as opposed to inner vectors. 

Changes were needed in JsonFileReader (function readFromJsonIntoVector) and JsonFileWriter (function writeFromVectorIntoJson) respectively.

On the C++ side, I believe similar changes are needed in json.cc (reader and writer implementation) and json_internal.cc (ReadRecordBatch and WriteRecordBatch) functions assuming I am looking at the correct Json handling code in C++.

@BryanCutler, @icexelloss , since you had reviewed the Json File reader/writer changes and later Li refactored them, do you agree with above analysis and suggested changes on the C++ side?

icexelloss commented:
@siddharthteotia There is already a bunch of style check warning. But there shouldn't be any with regard to Indentation, WhitespaceAround and EmptyLineSeparator.

wesm commented:
> It was changed to directly work with inner buffers as opposed to inner vectors.

Again, I'm pretty lost on this terminology. Which vector types are affected by this? Can you point me to the Java code in the old vectors vs. the new vectors so I can see for myself

icexelloss commented:
@wesm Decimal vectors are failing. I suspect this has something to do with recent changes on master with decimals.

wesm commented:
Got it, thanks. Unless the JSON format is changing, I don't see what could be wrong with the C++ implementation, unless something about the binary format itself changed

wesm commented:
cc @cpcloud

BryanCutler commented:
I don't see anything obvious wrong with the decimal reading here, I'll reproduce locally and take a closer look.

icexelloss commented:
I think I found the issue:

https://github.com/siddharthteotia/arrow/blob/ARROW-1476/java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java#L516

Comparing to the old code:
https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/BitVector.java#L57

The old code handles the case where nullCount == 0 and validity buffer is empty. The new code doesn't.

BryanCutler commented:
regarding whitespace style, I saw there were a few files with no empty line at the end.  Is that part of the style check?

siddharthteotia commented:
Thanks @icexelloss . It looks like the loadFieldBuffers code in new vector is missing some special handling of validity buffer. However, as per the old load code I am not sure when (and why) will the readable bytes be 0 and value count non-zero -- https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/BitVector.java#L57
If the condition is not fulfilled then it does what the new code is doing -- associate the buffer with target allocator. I am not quite sure when will the specialized handling be exercised? 

Whether the vector is all NULLs or not, the buffer cannot be empty if the value count is non-zero. Am I missing something?

icexelloss commented:
@siddharthteotia I think validity buffer is optional when nullCount == 0 per Arrow spec. I think C++ actually sends empty validity buffer when nullCount == 0.

BryanCutler commented:
Yeah, this is what's happening.  I confirmed the failure is when reading an binary file from C++ where null count is 0.

siddharthteotia commented:
Thanks @BryanCutler , @icexelloss . I wasn't aware of this and assumed that always a non-empty validity buffer will be shipped as long as valueCount > 0 regardless of null count.

icexelloss commented:
Created https://github.com/siddharthteotia/arrow/pull/2 to detect incorrect license headers.

BryanCutler commented:
BTW, I think we only saw this failure because the random seed in the integration tests led a dictionary record batch that sampled all valid bits.  Perhaps we ought to have an explicit integration test for this case?

icexelloss commented:
@BryanCutler I added module for "NewlineAtEndOfFile" here https://github.com/siddharthteotia/arrow/pull/2

siddharthteotia commented:
There was one more problem. When JsonFileReader was  refactored recently while removing the static interfaces, we did not set the writer index on the buffers we are allocating for the vector. As we are parsing JSON and writing to buffers, we should be setting writer index before calling loadFieldBuffers(). 

The static interfaces were doing so but we missed doing that in the recent change

BryanCutler commented:
FYI, I updated Spark to use these changes and tests pass for the most part.  There are still a couple things I need to work out, but overall it is looking good from this side!

icexelloss commented:
Oops, I can take a look at JsonFileReader tomorrow. @siddharthteotia which tests are failing because of that?

siddharthteotia commented:
Fixing checkstyle for warning is probably a wasted effort especially when the warnings are coming from code that hasn't been modified as well. I have filed a JIRA to address this separately. With so many warnings dumped by mvn package checkstyle:check, it is slightly difficult to identify what to fix and what not.

Essentially everything needs to be fixed -- indentation, order of imports etc throughout the code base. 

I am fine with doing it but can we do it separately? I have filed a JIRA to track this 
https://issues.apache.org/jira/browse/ARROW-1813

wesm commented:
Yes, let's do the style fixes separately. I think it's long overdue

icexelloss commented:
@siddharthteotia How about we fix at least indentation, header and white spaces, end of file new lines for new files? 

Other checkstyle warning we can fix it later, but I am concerned with adding lots of new files with incorrect indentation and white spaces.

siddharthteotia commented:
@icexelloss , I have fixed indentation, license header and white spaces in new files. Going through the code again. Travis CI build passes.

siddharthteotia commented:
Build passed.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/VectorUnloader.java:
```diff
@@ -27,6 +27,8 @@
 import org.apache.arrow.vector.schema.ArrowRecordBatch;
 import org.apache.arrow.vector.schema.ArrowVectorType;
 
+import javax.annotation.Nullable;
```
Unused?

siddharthteotia replied:
Done.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java:
```diff
@@ -218,16 +190,310 @@ public VectorSchemaRoot read() throws IOException {
+          buf.writeByte(parser.getByteValue());
+        }
+
+        buf.writerIndex(size);
```
Do we need this? `buf.writeByte` seems to update writerIndex already.

siddharthteotia replied:
Yes we don't need.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java:
```diff
@@ -363,86 +481,114 @@ public UnionVector promoteToUnion() {
-      lastSet = index + 1;
-      return offsets.getAccessor().get(lastSet);
-    }
+  public int getValidityAndOffsetValueCapacity() {
```
@siddharthteotia I think we want to make this private, is that correct?

siddharthteotia replied:
Yes.

icexelloss commented:
Awesome.

@siddharthteotia I have re commented on all my previous comments that are not addressed. Can you take a look? Thanks!

icexelloss commented:
@siddharthteotia There are still some indentation and white space warning on the new vector files, you can see this by running:

```
mvn package checkstyle:checkstyle
```

and check the html file under 

```
vector/target/site/checkstyle.html
```

Can you double check those please? Thanks

wesm commented:
Looks like we are on the home stretch. I can merge later today when these last small items are addressed

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -0,0 +1,848 @@
+   *         vector instance.
+   */
+  @Override
+  public ArrowBuf[] getBuffers(boolean clear) {
```
My previous comments are squashed because of code change.

Why do we need both 

```
public ArrowBuf[] getBuffers(boolean clear)
``` 
and 
```
public List<ArrowBuf> getFieldBuffers()
```
Both as public API? They look very similar to me and might confuse users.

siddharthteotia replied:
AFAIK, getBuffers() is an old method for getting a batch of buffers for serialization. We use it for maintaining compatibility with Drill client. I think in future, we should consider removing it from Arrow and writing a wrapper over it in Dremio. This way we will just have getFieldBuffers.

icexelloss replied:
I see. Does it make sense to mark getBuffers() as deprecated so it's less confusing for the other users?

siddharthteotia replied:
Let's not worry about it right now. These two APIs have not been introduced as part of refactoring and have been in Arrow code for quite some time. 

I will file a JIRA and we can brainstorm w.r.t appropriate steps to address the ambiguity (if any)

icexelloss replied:
Ok. Sounds good.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -0,0 +1,1250 @@
+   * @param index   position of the element to set
+   * @param length  length of the element
+   */
+  public void setValueLengthSafe(int index, int length) {
```
This function is not used. 

Also, I am not sure if we need an API for just setting value length. When would this be useful?

icexelloss replied:
@siddharthteotia what's your thought on this

siddharthteotia replied:
We use it in our Parquet code -- I may have to evaluate its usage but the API has been there in variable width vector from the beginning.

icexelloss replied:
Thanks for the clarification - I wasn't sure which APIs are from the old vectors and which APIs are added.

Looks good then.

siddharthteotia commented:
@wesm , let me know if I need to squash the latest 7 commits for this JIRA or you can take care of it while merging. I think we would be merging 4 commits into master:

ARROW-1473
ARROW-1474
ARROW-1717
ARROW-1476 (squashed latest commits)

siddharthteotia commented:
Build passed.
I think we are all set for merge.

icexelloss commented:
LGTM. +1. 

Thanks @siddharthteotia for the epic effort!

wesm commented:
Sweet, thanks all! I will take care of merging to master (and squashing the commits between the last refactor patch and now)

BryanCutler commented:
+1, great work @siddharthteotia !

wesm commented:
merged to master. thank you!
